src/ucp/core/ucp_mm.c

```diff
@@ -803,21 +798,25 @@ ucp_memh_init_from_parent(ucp_mem_h memh, ucp_md_map_t parent_md_map)
     }
 }
 
-static ucs_status_t
-ucp_memh_init_uct_reg(ucp_context_h context, ucp_mem_h memh,
-                      ucp_md_map_t reg_md_map, unsigned uct_flags,
-                      const char *alloc_name)
+static ucs_status_t ucp_memh_init_uct_reg(ucp_context_h context, ucp_mem_h memh,
+                                          unsigned uct_flags,
+                                          const char *alloc_name)
 {
     ucs_memory_type_t mem_type = memh->mem_type;
+    ucp_md_map_t reg_md_map    = context->reg_md_map[mem_type];
     void *address              = ucp_memh_address(memh);
     size_t length              = ucp_memh_length(memh);
     ucp_md_map_t cache_md_map;
     ucs_status_t status;
 
+    if (uct_flags & UCT_MD_MEM_FLAG_LOCK) {
+        reg_md_map |= context->reg_block_md_map[mem_type];
+    }
+
+    reg_md_map  &= ~memh->md_map;
     cache_md_map = context->cache_md_map[mem_type] & reg_md_map;
 
-    if ((context->rcache == NULL) || (cache_md_map == 0) ||
-        (memh->flags & UCP_MEMH_FLAG_NO_RCACHE)) {
+    if ((context->rcache == NULL) || (memh->flags & UCP_MEMH_FLAG_NO_RCACHE)) {
```

Comment:
why can we remove check for `cache_map_md == 0`?

Reply:
oh right, i meant to remove "(context->rcache == NULL)" check in the previous comment, not (cache_map_md == 0)

Reply:
In [the previous commit](https://github.com/openucx/ucx/pull/10053), I added a call to `ucp_memh_init_uct_reg()` from `ucp_mem_type_reg_buffers()`. The `cache_map_md == 0` check was introduced to allow gdr_copy (or other non-cacheable MD) to bypass the UCP rcache lookup and proceed directly to registration. In this PR, I rewrite the logic, and since the call to `ucp_memh_init_uct_reg()` is removed, I am reverting this change as well.